django/db/models/sql/query.py

```diff
@@ -2624,14 +2631,9 @@ def is_nullable(self, field):
         nullable for those backends. In such situations field.null can be
         False even if we should treat the field as nullable.
         """
-        # We need to use DEFAULT_DB_ALIAS here, as QuerySet does not have
-        # (nor should it have) knowledge of which connection is going to be
```

Comment:
I think [this comment nails it](https://code.djangoproject.com/ticket/25947#comment:1), _teaching_ `sql.Query` about db alias seems wrong as this abstraction was designed to be backend agnostic.

Moving forward with this change gives the impression that `self.db` will always be the right choice over `DEFAULT_DB_ALIAS` but in most cases the proper database to use won't be known until execution time as it will be determine dynamically from `django.db.router`. All this code does is _fix_ the crash when `using()` explicitly used.

The impact of the changes proposed here go well beyond fixing what was brought up in ticket-25947 which is that `Query.__str__()` would crash if there was no `DATABASES["default"]` alias defined. If we really want to prevent that from happening (lots of things are going to break if you don't set `DATABASES["default"]` btw) I suggest we define a property that does

```python
@cached_property
def _default_db_alias(self):
    if DEFAULT_DB_ALIAS in connections:
        return DEFAULT_DB_ALIAS
    # Pick the first defined alias if "default" is mising
    return next(iter(connections), DEFAULT_DB_ALIAS)
```

Trying to solve this particular use case of supporting a lack of `DATABASES["default"]` is a slippery slope though as tens of features expect it to be defined. Addressing this issue sends a signal that this is a blessed configuration and that we commit to solving other `DEFAULT_DB_ALIAS` usages.

---

By looking at this ticket and [ticket-25705](https://code.djangoproject.com/ticket/25705#comment:12) recently I'm starting to believe that we should combine these two tickets into a single one that keep things as it for `Query.__str__()` and document `Query.sql_with_params(using: str = DEFAULT_DB_ALIAS)` instead.

That would allow users that have exotic `DATABASES` configurations to use this method instead to target a particular aliases, avoid venturing down the path of safe and correct parameters escaping for all backends (and force third party ones to do so as well), enforce the safe usages of `QuerySet.raw`.

Reply:
How about also set Query.db in QuerySet.__init__ ? This will fix the crash when default db is empty .

Reply:
OK, thank you for all the explanation.